### PR TITLE
[CALCITE-6741] The type of a comparison is nullable when either opera…

### PR DESCRIPTION
**Pull request by mihaibudiu**
…nd is nullable

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 1]
```diff
@@ -4480,7 +4480,7 @@ private void checkNegWindow(String s, String msg) {
     expr("(1,2) in ((1,2), (3,4))")
         .columnType("BOOLEAN NOT NULL");
     expr("'medium' in (cast(null as varchar(10)), 'bc')")
-        .columnType("BOOLEAN NOT NULL");
```
Comment (NobiGo) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:4483]: In my opinion, this change is reasonable, but it's strange that it only affects this test case. Do we need to add more test cases to cover the corresponding changes?

Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:4483]: I suspect is affects many programs, but there a few reasons we don't see this:
- this affects only some subexpressions, but often the final expression in a Project has the same type
- many program representations do not display the types of subexpressions
- the Enumerable representation is in Java, where all Objects are nullable
- there aren't enough tests with complex expressions and null values
Our runtime is written in Rust, and there are separate types for non-nullable (T) and nullable (Option<T>). Many mistakes are caught at compile-time. Other mistakes that are not caught at compile-time become panics at runtime due to unwrap() calls on Option objects. In Java both of these may be silent.
It's not easy to write good tests to catch these problems. I think adding more tests like the one that has failed has value, but only limited value.

Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:4483]: I get it. If there are no test cases, future PRs are likely to reverse the current changes.

Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:4483]: @NobiGo see a second commit with more coercion tests

**Review by NobiGo — APPROVED** [review]
LGTM

**Pull request merged**